compiler/x/codegen/OMRCodeGenerator.cpp

```diff
@@ -229,19 +233,24 @@ OMR::X86::CodeGenerator::initialize(TR::Compilation *comp)
    //
 
 #if defined(TR_TARGET_X86) && !defined(J9HAMMER)
-   if (_targetProcessorInfo.supportsSSE2() && comp->target().cpu.testOSForSSESupport())
+   TR_ASSERT(comp->target().cpu.supportsFeature(OMR_FEATURE_X86_SSE2) == _targetProcessorInfo.supportsSSE2(), "supportsSSE2() failed\n");
+   
+   if (comp->target().cpu.supportsFeature(OMR_FEATURE_X86_SSE2) && comp->target().cpu.testOSForSSESupport())
       supportsSSE2 = true;
 #endif // defined(TR_TARGET_X86) && !defined(J9HAMMER)
 
-   if (_targetProcessorInfo.supportsTM() && !comp->getOption(TR_DisableTM))
+   TR_ASSERT(comp->target().cpu.supportsFeature(OMR_FEATURE_X86_RTM) == _targetProcessorInfo.supportsTM(), "supportsTM() failed\n");
+
+   if (comp->target().cpu.supportsFeature(OMR_FEATURE_X86_RTM) && !comp->getOption(TR_DisableTM))
```

Comment:
Never been a fan of this option checking along with processor feature support. Why don't we just disable the feature at the compiler level once and not have to check for `TR_DisableTM` everywhere we check for the feature?

Reply:
We initialize TR::Compiler before the options get initialized (Both OpenJ9 and JITBuilder seems to do this).
I don't think this will be as simple as checking the options during processor detection initialization and disable.
I was forced to have a workaround on Z because of this issue.
I will create that Z PR now (still in my local repo) and if you are okay with the workaround for now then i'll apply it to this one as well.
And if not, we might wanna deal with this issue in a separate PR.

Reply:
Sure I'm ok with that approach.

Reply:
I uploaded the Z PR and pointed you to the workaround (https://github.com/eclipse/openj9/pull/9431).

Reply:
@harryyu1994 I took a look at the PR you pointed to. I still don't quite understand what the restriction is. Sure, we have to call that `applyUserOptions` API at some later point when options are available, however that `applyUserOptions` should be the one that should check in this case whether `TR_DisableTM` option is active, and if so then to turn off the facility at the OMRCPU level. 

In fact I don't believe that `applyUserOptions` API belongs in the CPU class. The CPU class should not have to deal with options, etc. That is an external users prerogative. However that is another discussion all together.

The fact still remains here that we shouldn't need to check the facility and the option together all the time. Some previous code will have checked the option and turned off the facility for us.

Reply:
So if you are fine with me introducing this `applyUserOptions` for x to deal with options then I will add it in so that we won't check facility and option together all the time. There's no restriction if we are allowed to have something like an `applyUserOptions`. The only restriction is we can't do option checking in `detect()`.

Reply:
Also I feel like there's enough changes in this PR and that can be done in a separate PR.

Reply:
I agree, we can leave this off for another PR. Doing my best to ensure all platforms are on equal footing moving forward and we don't do special stuff on one platform vs. others. The more we consolidate and keep things the same the easier it will be to reason about. Thanks!

Reply:
Issue opened to track leftover (https://github.com/eclipse/omr/issues/5247)